memory/alloydb/chat_message_history.go

```diff
@@ -0,0 +1,180 @@
+package alloydb
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/tmc/langchaingo/internal/alloydbutil"
+	"github.com/tmc/langchaingo/llms"
+	"github.com/tmc/langchaingo/schema"
+)
+
+type ChatMessageHistory struct {
+	engine     alloydbutil.PostgresEngine
+	sessionID  string
+	tableName  string
+	schemaName string
+	overwrite  bool
+}
+
+var _ schema.ChatMessageHistory = &ChatMessageHistory{}
+
+// NewChatMessageHistory creates a new NewChatMessageHistory with options.
+func NewChatMessageHistory(ctx context.Context, engine alloydbutil.PostgresEngine, tableName string, opts ...ChatMessageHistoryStoresOption) (ChatMessageHistory, error) {
+	cmh, err := ApplyChatMessageHistoryOptions(engine, tableName, opts...)
+	if err != nil {
+		return ChatMessageHistory{}, err
+	}
+	err = cmh.validateTable(ctx)
+	if err != nil {
+		return ChatMessageHistory{}, err
+	}
+	return cmh, nil
+}
+
+// validateTable validates if a table with a specific schema exist and it
+// contains the required columns.
+func (c *ChatMessageHistory) validateTable(ctx context.Context) error {
+	tableExistsQuery := fmt.Sprintf(`SELECT EXISTS (
+		SELECT FROM information_schema.tables 
+		WHERE table_schema = '%s' AND table_name = '%s');`, // TODO :: Are table/schema names which require case-sensitive?
+		c.schemaName, c.tableName)
+	var exists bool
+	err := c.engine.Pool.QueryRow(ctx, tableExistsQuery).Scan(&exists)
+	if err != nil {
+		return fmt.Errorf("error validating table %s: %w", c.tableName, err)
+	}
+	if !exists {
+		return fmt.Errorf("table '%s' does not exist in schema '%s'", c.tableName, c.schemaName)
+	}
+
+	requiredColumns := []string{"id", "session_id", "data", "type"} // TODO :: Should the field be data or content?
+	for _, reqColumn := range requiredColumns {
+		columnExistsQuery := fmt.Sprintf(`SELECT EXISTS (
+			SELECT 1 FROM information_schema.columns 
+			WHERE table_schema = '%s' AND table_name = '%s' AND column_name = '%s'
+		);`, c.schemaName, c.tableName, reqColumn)
+		var columnExists bool
+		err := c.engine.Pool.QueryRow(ctx, columnExistsQuery).Scan(&columnExists)
+		if err != nil {
+			return fmt.Errorf("error scanning columns from table %s: %w", c.tableName, err)
+		}
+		if !columnExists {
+			return fmt.Errorf("column '%s' is missing in table '%s'. Expected columns: %v", reqColumn, c.tableName, requiredColumns)
+		}
+	}
+	return nil
+}
+
+// addMessage inserts a new message into AlloyDB.
+func (c *ChatMessageHistory) addMessage(_ context.Context, content string, messageType llms.ChatMessageType) error {
+	query := fmt.Sprintf(`INSERT INTO "%s"."%s" (session_id, data, type) VALUES ($1, $2, $3)`,
+		c.schemaName, c.tableName)
+
+	_, err := c.engine.Pool.Exec(context.Background(), query, c.sessionID, content, messageType)
+	if err != nil {
+		return fmt.Errorf("failed to add message to database: %w", err)
+	}
+	return nil
+}
+
+// AddMessage adds a message to the chat message history.
```

Comment:
nit: (please apply elsewhere as well)
```suggestion
// AddMessage adds a message to the chat message history.
```